docs/configuration.md
# Configuration

You can use environment variables to set configuration values.

## Security

| Environment variable        | Description                     | Default value |
| --------------------------- | ------------------------------- | ------------- |
| `SALT_ROUNDS`               | Number of password salt rounds  | 10            |
| `JWT_SECRET`                | Secret to sign JWTs             | Prep-AI       |
| `TOTP_WINDOW_PAST`          | Expired TOTP tokens to accept   | 1             |
| `TOTP_WINDOW_FUTURE`        | Future TOTP tokens to accept    | 0             |
| `MFA_TOKEN_EXPIRY`          | Expiry time for MFA tokens      | 10m           |
| `MERGE_USERS_TOKEN_EXPIRY`  | Expiry time to merging link     | 30m           |
| `ACCESS_TOKEN_EXPIRY`       | Expiry time for access tokens   | 1h            |
| `PASSWORD_PWNED_CHECK`      | Check for Pwned passwords       | false         |
| `DELETE_EXPIRED_SESSIONS`   | Delete inactive sessions (days) | 30            |
| `INACTIVE_USER_DELETE_DAYS` | Delete deactivated users (days) | 30            |

## Email

You can set the following environment variables to specify the name and from email address:

| Environment variable | Description         |
| -------------------- | ------------------- |
| `EMAIL_NAME`         | Name of the service |
| `EMAIL_FROM`         | From email address  |

If you want to use SMTP, you should additionally set the configuration:

| Environment variable | Description      |
| -------------------- | ---------------- |
| `EMAIL_HOST`         | Host             |
| `EMAIL_PORT`         | Port             |
| `EMAIL_SECURE`       | Secure (boolean) |
| `EMAIL_USER`         | Username         |
| `EMAIL_PASSWORD`     | Password         |

Alternately, if you want to use AWS SES, you should set these instead (note that you can also use SMTP with SES):

| Environment variable          | Description    |
| ----------------------------- | -------------- |
| `EMAIL_SES_ACCESS_KEY_ID`     | AWS access key |
| `EMAIL_SES_SECRET_ACCESS_KEY` | AWS secret key |
| `EMAIL_SES_REGION`            | AWS region     |

To generate an access/secret key pair, you can create an IAM user with the permission `AmazonSESFullAccess`. For more details, read the article [Creating an IAM user in your AWS account](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_create.html#id_users_create_console) on the AWS website.

### Rate limiting

Prep-AI API has three types of rate limits. When an endpoint is accessed, 1 point is consumed. There are also some endpoints that consume additional points (like logging in or creating an account consumes 10 points). The types of rate limits are:

1. "Public" for unauthenticated requests (250 points/hour)
2. "Authenticated" for requests with a user access token (5k points/hour)
3. "API key" for (automated) requests using an API key (10k points/hour)

You can set the rate limits for each of these categories. By default, the rate limit resets after one hour:

| Environment variable                | Description                      | Default |
| ----------------------------------- | -------------------------------- | ------- |
| `RATE_LIMIT_PUBLIC_POINTS`          | Maximum points for public        | 250     |
| `RATE_LIMIT_PUBLIC_DURATION`        | Reset duration for public        | 3600    |
| `RATE_LIMIT_AUTHENTICATED_POINTS`   | Maximum points for authenticated | 5000    |
| `RATE_LIMIT_AUTHENTICATED_DURATION` | Reset duration for authenticated | 3600    |
| `RATE_LIMIT_API_KEY_POINTS`         | Maximum points for API key       | 10000   |
| `RATE_LIMIT_API_KEY_DURATION`       | Reset duration for API key       | 3600    |

## Optional services

### ElasticSearch

ElasticSearch is used for tracking API key logs.

If you have a public ElasticSearch instance (this is not recommended), you only need to specify the node:

```env
ELASTICSEARCH_NODE = "https://your-endpoint.example"
```

If your endpoint uses HTTP basic authentication, you can add the credentials:

```env
ELASTICSEARCH_AUTH_USERNAME = "Your username"
ELASTICSEARCH_AUTH_PASSWORD = "Your password"
```

Or, if you're using an ElasticSearch-hosted instance with an API key, you can provide only the API key or a combination of the API key and ID:

```env
ELASTICSEARCH_AUTH_API_KEY = "Your API key"
ELASTICSEARCH_AUTH_API_KEY_ID = "Your API key ID"
```

Alternately, if you're using the Amazon Elasticsearch Service, you can specify the AWS credentials:

```env
ELASTICSEARCH_NODE = "https://search-your-endpoint.us-east-1.es.amazonaws.com"
ELASTICSEARCH_AWS_ACCESS_KEY_ID = "Your AWS access key ID"
ELASTICSEARCH_AWS_SECRET_ACCESS_KEY = "Your AWS secret access key"
ELASTICSEARCH_AWS_REGION = "us-east-1"
```

### Twilio SMS

To send SMS messages using Twilio, you should set the following environment variables:

| Environment variable | Description           |
| -------------------- | --------------------- |
| `TWILIO_ACCOUNT_SID` | Twilio account SID    |
| `TWILIO_AUTH_TOKEN`  | Twilio auth token     |
| `SMS_FAIL_RETRIES`   | Number of SMS retries |


docs/environment-variables.md
# Eenvironment variables

All environment variables are stored in a `.env` file locally. All environment variables are optional, with the exception of the database URL.

## Port

The first and most important is the port you want the PrepAI backend process to run on. On production, this is usually `80`. Locally, it can be something like `3000` or `8080`. In these examples, we'll use the port 7007.

```env
PORT = 7007
```

This will launch the app on http://localhost:7007.

### URLs

When sending emails and performing redirects, the URL of your frontend app is required. In most cases, this is the URL on which [Prep-AI UI](https://github.com/Prep-AI/ui) is running:

```env
FRONTEND_URL = "https://example.com"
```

### Database connection

Once you have your database details (host, username, password, and databse name), you can create a file with the path `prisma/.env` with your database URL:

```env
DATABASE_URL = "postgresql://USER:PASSWORD@HOST:PORT/DATABASE?schema=SCHEMA"
```

Similarly, a MySQL or MariaDB URL looks like:

```env
DATABASE_URL = "mysql://USER:PASSWORD@HOST:PORT/DATABSE"
```

You can read more about this URL on the [Add to existing project](https://www.prisma.io/docs/getting-started/setup-prisma/add-to-existing-project-typescript-postgres#connect-your-database) article on the Prisma docs website.

### Sending emails

Prep-AI API sends transactions emails for email verification, password resets, etc. You can either send emails using SMTP or AWS SES. If you want to use SMTP:

```env
EMAIL_HOST = "smtp.example.com"
EMAIL_FROM = "hello@example.com"
EMAIL_PASSWORD = "your-password"
```

Alternately, you can use AWS SES to send emails:

```env
SES_EMAIL = "hello@example.com"
SES_REGION = "eu-west-2"
SES_ACCESS = "aws-access-key-xxxxxxxxxx"
SES_SECRET = "aws-secret-key-xxxxxxxxxx"
```

In the above example, emails are sent from `hello@example.com` and SES is setup in the `eu-west-2` AWS region with the given credentials.

### Billing

We use Stripe for billing and subscription management. You'll need your Stripe secret key and a product ID (you can use your test key to make sure people aren't actually charged). We use the product ID to show the pricing plans of that product.

```env
STRIPE_SECRET_KEY = "stripe-test-api-key"
STRIPE_PRODUCT_ID = "stripe-product-id"
```

### Encryption keys

It is highly recommended that you change the default encryption keys for signing JWTs, salts for hashes, etc. These should just be large (10+ character) strings, preferably a mix of letters, numbers, and special characters:

```env
JWT_SECRET = "secret"
TWT_SECRET = "secret"
```

### Redis

Prep-AI API uses both Redis-based and in-memory caches to optimize queries and requests. If you’re running Redis using the default settings, you can ignore the `REDIS_URL` key, or populate it with your instance’s URL if it’s managed or on another port:

The TTL determines how long a result is cached in the memory; by default, this is 10 minutes. The check period tells Prep-AI how often to check the cache and remove expired items (think of it like a `setInterval`). Organization details, API key scopes, etc., are stored in-memory for faster access, and Redis is primarily used to invalidate JWTs.

```env
REDIS_URL = "redis://127.0.0.1:6379"
CACHE_TTL = 600                  #     10 mins
CACHE_CHECK_PERIOD = 1000        #     1,000 s
```

### Rate limits

Rate limits help Prep-AI mitigate some types of attacks. By default, users can do 60 requests/minute without an API key. With API key authentication, this is increased to 1,000 requests/minute. There is also a speed limit mechanism that delays requests by 100ms (in response time) per request if more than 500 are made in a minute.

Brute force prevention is used on authentication endpoints, where only 50 requests per 5 minutes are allowed, and users have to wait for (an increasing amount of) some time before they can make another request.

```env
## Brute force is used for auth endpoints
BRUTE_FREE_RETRIES = 50          # 50 requests
BRUTE_LIFETIME = 300000          #   in 5 mins

## Public limits
PUBLIC_RATE_LIMIT_MAX = 60       # 60 requests
PUBLIC_RATE_LIMIT_TIME = 60000   #    in 1 min
SPEED_LIMIT_COUNT = 500         # 1k requests
SPEED_LIMIT_TIME = 600000        #    in 1 min
SPEED_LIMIT_DELAY = 100          # delay 100ms

## Limits when using an API key
RATE_LIMIT_MAX = 1000            # 1k requests
RATE_LIMIT_TIME = 60000          #    in 1 min
```

### Expiry durations

JWTs are used for authentication and emails, and you can configure their expiry duration. Email verification links are valid for one week, password reset links for a day, and location approval links for 10 minutes.

In terms of authentication, a login session token is approved for 15 minutes and a refresh token is valid for 1 month.

```env
TOKEN_EXPIRY_EMAIL_VERIFICATION = "7d"
TOKEN_EXPIRY_PASSWORD_RESET = "1d"
TOKEN_EXPIRY_LOGIN = "15m"
TOKEN_EXPIRY_APPROVE_LOCATION = "10m"
TOKEN_EXPIRY_REFRESH = "30d"
```

### Disallow disposable emails

If you don't want users to be able to sign up using a disposable email address, set this to `false`.

```env
ALLOW_DISPOSABLE_EMAILS = false
```

### ElasticSearch for data and tracking

We use ElasticSearch to track events, analytics, and server logs. You can use AWS-managed ElasticSearch or your own instance. If you use AWS:

```env
AWS_ELASTIC_HOST = "https://name.region.es.amazonaws.com"
AWS_ACCESS_KEY_ID = "aws-access-key-xxxxxxxxxx"
AWS_SECRET_ACCESS_KEY = "aws-secret-key-xxxxxxxxxx"
AWS_REGION = "eu-west-2"
```

Alternately, you can also use a custom ElasticSearch service if you don’t want to use AWS. In this case, you have to specify the host, log, and API version.

### Sentry error tracking

If you want to track errors using Sentry, you can supply your DSN:

```env
SENTRY_DSN = "https://<key>@sentry.io/<project>"
```


src/config/configuration.ts
import { ConfigFactory } from '@nestjs/config/dist/interfaces';
import { config } from 'dotenv';
import { expand } from 'dotenv-expand';
import { Configuration } from './configuration.interface';

const envConfig = config();
expand(envConfig);

const int = (val: string | undefined, num: number): number =>
  val ? (isNaN(parseInt(val)) ? num : parseInt(val)) : num;
const bool = (val: string | undefined, bool: boolean): boolean =>
  val == null ? bool : val == 'true';

export default () => ({
  DATABASE_URL: process.env.DATABASE_URL,
  port: parseInt(process.env.PORT, 10) || 3000,
  database: {
    host: process.env.DATABASE_HOST,
    port: parseInt(process.env.DATABASE_PORT, 10) || 5432,
  },
  frontendUrl: process.env.FRONTEND_URL ?? 'http://localhost:3000',
  meta: {
    appName: process.env.APP_NAME ?? 'prepai',
    domainVerificationFile:
      process.env.DOMAIN_VERIFICATION_FILE ?? 'prepai-verify.txt',
  },
  rateLimit: {
    public: {
      points: int(process.env.RATE_LIMIT_PUBLIC_POINTS, 250),
      duration: int(process.env.RATE_LIMIT_PUBLIC_DURATION, 3600),
    },
    authenticated: {
      points: int(process.env.RATE_LIMIT_AUTHENTICATED_POINTS, 5000),
      duration: int(process.env.RATE_LIMIT_AUTHENTICATED_DURATION, 3600),
    },
    apiKey: {
      points: int(process.env.RATE_LIMIT_API_KEY_POINTS, 10000),
      duration: int(process.env.RATE_LIMIT_API_KEY_DURATION, 3600),
    },
  },
  caching: {
    geolocationLruSize: int(process.env.GEOLOCATION_LRU_SIZE, 100),
    apiKeyLruSize: int(process.env.API_KEY_LRU_SIZE, 100),
  },
  security: {
    saltRounds: int(process.env.SALT_ROUNDS, 10),
    jwtSecret: process.env.JWT_SECRET ?? 'prepai',
    totpWindowPast: int(process.env.TOTP_WINDOW_PAST, 1),
    totpWindowFuture: int(process.env.TOTP_WINDOW_FUTURE, 0),
    mfaTokenExpiry: process.env.MFA_TOKEN_EXPIRY ?? '10m',
    mergeUsersTokenExpiry: process.env.MERGE_USERS_TOKEN_EXPIRY ?? '30m',
    accessTokenExpiry: process.env.ACCESS_TOKEN_EXPIRY ?? '1h',
    passwordPwnedCheck: bool(process.env.PASSWORD_PWNED_CHECK, true),
    unusedRefreshTokenExpiryDays: int(process.env.DELETE_EXPIRED_SESSIONS, 30),
    inactiveUserDeleteDays: int(process.env.INACTIVE_USER_DELETE_DAYS, 30),
  },
  elasticSearch: {
    node: process.env.ELASTICSEARCH_NODE,
    retries: int(process.env.ELASTICSEARCH_FAIL_RETRIES, 3),
    auth: process.env.ELASTICSEARCH_AUTH_USERNAME
      ? {
          username: process.env.ELASTICSEARCH_AUTH_USERNAME,
          password: process.env.ELASTICSEARCH_AUTH_PASSWORD,
        }
      : process.env.ELASTICSEARCH_AUTH_API_KEY
        ? process.env.ELASTICSEARCH_AUTH_API_KEY_ID
          ? {
              apiKey: {
                api_key: process.env.ELASTICSEARCH_AUTH_API_KEY,
                id: process.env.ELASTICSEARCH_AUTH_API_KEY_ID,
              },
            }
          : { apiKey: process.env.ELASTICSEARCH_AUTH_API_KEY }
        : undefined,
    aws: {
      accessKeyId: process.env.ELASTICSEARCH_AWS_ACCESS_KEY_ID ?? '',
      secretAccessKey: process.env.ELASTICSEARCH_AWS_SECRET_ACCESS_KEY ?? '',
      region: process.env.ELASTICSEARCH_AWS_REGION ?? '',
    },
  },
  email: {
    name: process.env.EMAIL_NAME ?? 'prepai',
    from: process.env.EMAIL_FROM ?? '',
    retries: int(process.env.EMAIL_FAIL_RETRIES, 3),
    ses: {
      accessKeyId: process.env.EMAIL_SES_ACCESS_KEY_ID ?? '',
      secretAccessKey: process.env.EMAIL_SES_SECRET_ACCESS_KEY ?? '',
      region: process.env.EMAIL_SES_REGION ?? '',
    },
    transport: {
      host: process.env.EMAIL_HOST ?? '',
      port: int(process.env.EMAIL_PORT, 587),
      secure: bool(process.env.EMAIL_SECURE, false),
      auth: {
        user: process.env.EMAIL_USER ?? process.env.EMAIL_FROM ?? '',
        pass: process.env.EMAIL_PASSWORD ?? '',
      },
    },
  },
  webhooks: {
    retries: int(process.env.WEBHOOK_FAIL_RETRIES, 3),
  },
  sms: {
    retries: int(process.env.SMS_FAIL_RETRIES, 3),
    twilioAccountSid: process.env.TWILIO_ACCOUNT_SID ?? '',
    twilioAuthToken: process.env.TWILIO_AUTH_TOKEN ?? '',
  },
  payments: {
    stripeApiKey: process.env.STRIPE_API_KEY ?? '',
    stripeProductId: process.env.STRIPE_PRODUCT_ID ?? '',
    stripeEndpointSecret: process.env.STRIPE_ENDPOINT_SECRET ?? '',
    paymentMethodTypes: ['card'],
  },
  tracking: {
    mode:
      (process.env.TRACKING_MODE as Configuration['tracking']['mode']) ??
      'api-key',
    index: process.env.TRACKING_INDEX ?? 'prepai-logs',
    deleteOldLogs: bool(process.env.TRACKING_DELETE_OLD_LOGS, true),
    deleteOldLogsDays: int(process.env.TRACKING_DELETE_OLD_LOGS_DAYS, 90),
  },
  slack: {
    token: process.env.SLACK_TOKEN ?? '',
    slackApiUrl: process.env.SLACK_API_URL,
    rejectRateLimitedCalls: bool(
      process.env.SLACK_REJECT_RATE_LIMITED_CALLS,
      false,
    ),
    retries: int(process.env.SLACK_FAIL_RETRIES, 3),
  },
  s3: {
    accessKeyId: process.env.AWS_S3_ACCESS_KEY ?? '',
    secretAccessKey: process.env.AWS_S3_SECRET_KEY ?? '',
    region: process.env.AWS_S3_REGION ?? '',
    profilePictureBucket: process.env.AWS_S3_PROFILE_PICTURE_BUCKET ?? '',
    profilePictureCdnHostname:
      process.env.AWS_S3_PROFILE_PICTURE_CDN_HOST_NAME ?? '',
  },
  cloudinary: {
    cloudName: process.env.CLOUDINARY_CLOUD_NAME ?? '',
    apiKey: process.env.CLOUDINARY_API_KEY ?? '',
    apiSecret: process.env.CLOUDINARY_API_SECRET ?? '',
  },
  github: {
    auth: process.env.GITHUB_AUTH,
    userAgent: process.env.GITHUB_USER_AGENT,
  },
  googleMaps: {
    apiKey: process.env.GOOGLE_MAPS_API_KEY,
  },
  gravatar: {
    enabled: bool(process.env.PASSWORD_PWNED_CHECK, true),
  },
});

/* const configuration: Configuration = {
  
};
const configFunction: ConfigFactory<Configuration> = () => configuration;
export default configFunction; */
